zkevm-circuits/src/evm_circuit/util/memory_gadget.rs

```diff
@@ -173,23 +290,32 @@ impl<F: FieldExt, const MAX_QUAD_COST_IN_BYTES: usize>
         &self,
         region: &mut Region<'_, F>,
         offset: usize,
-        curr_memory_size: MemorySize,
-        address: Address,
-    ) -> Result<(MemorySize, u128), Error> {
+        curr_memory_size: u32,
+        addresses: [u64; N],
+    ) -> Result<(u64, u64), Error> {
         // Calculate the active memory size
-        let address_memory_size =
-            self.address_memory_size.assign(region, offset, address)?;
+        let memory_sizes = self
+            .memory_sizes
+            .iter()
+            .zip(addresses.iter())
+            .map(|(memory_size, address)| {
+                memory_size.assign(region, offset, *address)
+            })
+            .collect::<Result<Vec<_>, _>>()?;
 
         // Calculate the next memory size
-        let next_memory_size = self
-            .next_memory_size
-            .assign(
+        let mut next_memory_size = curr_memory_size as u64;
+        for (max_memory_sizes, memory_size) in
+            self.max_memory_sizes.iter().zip(memory_sizes.iter())
+        {
+            let (_, max) = max_memory_sizes.assign(
                 region,
                 offset,
-                F::from(address_memory_size),
-                F::from(curr_memory_size),
-            )?
-            .get_lower_128() as MemorySize;
+                F::from(next_memory_size as u64),
+                F::from(*memory_size),
+            )?;
+            next_memory_size = max.get_lower_128() as u64;
```

Comment:
Another conversion which could in theory fail no? Is this checked previously somewhere?

Reply:
Not really, let me try to review all and propagate the `Result` out.

Reply:
Ah, just realized the `lhs` and `rhs` or `MinMaxGadget` are both `u64`, so the conversion should not fail here.

Reply:
Ohhh that's right! My bad!!